### PR TITLE
fix: [BUG] processing block is skipped when web3 service is not available during Indexer execution

### PR DESCRIPTION
**Pull request by purplesmoke05**
**draft**
related to #1137.

### CODE REVIEW

File: batch/indexer_Position_Bond.py [hunk 1]
```diff
@@ -90,12 +93,18 @@ def initial_sync(self):
                     block_to=self.latest_block
                 )
             local_session.commit()
+        except Exception as e:
+            LOG.exception("An exception occurred during event synchronization")
+            local_session.rollback()
+            self.latest_block = latest_block_at_start
+            raise e
```
Comment (purplesmoke05) [inline batch/indexer_Position_Bond.py:100]: Rollback DB and instance value when processor catch an exception to raise.

File: batch/indexer_Position_Bond.py [hunk 2]
```diff
@@ -108,6 +117,11 @@ def sync_new_logs(self):
             )
             self.latest_block = blockTo
             local_session.commit()
+        except Exception as e:
+            LOG.exception("An exception occurred during event synchronization")
+            local_session.rollback()
+            self.latest_block = latest_block_at_start
+            raise e
```
Comment (purplesmoke05) [inline batch/indexer_Position_Bond.py:124]: Rollback DB and instance value when processor catch an exception to raise.

File: app/contracts/contract.py [hunk 3]
```diff
@@ -107,11 +107,12 @@ def call_function(contract: contract,
...
         try:
+            _function = getattr(contract.functions, function_name)
             result = _function(*args).call()
-        except BadFunctionCallOutput:
+        except (BadFunctionCallOutput, ABIFunctionNotFound) as web3_exception:
             if default_returns is not None:
                 return default_returns
             else:
-                raise BadFunctionCallOutput
+                raise web3_exception
```
Comment (purplesmoke05) [inline app/contracts/contract.py:116]: Fix handling error in web3 wrapper with same way as ibet-Prime.
https://github.com/BoostryJP/ibet-Prime/blob/90f849bbdacce9b8d8600552c134785917a37d19/app/utils/contract_utils.py#L138-L159

File: batch/indexer_Position_Bond.py [hunk 4]
```diff
@@ -176,6 +190,9 @@ def __sync_transfer(self, db_session: Session, block_from: int, block_to: int):
                     fromBlock=block_from,
                     toBlock=block_to
                 )
+            except ABIEventFunctionNotFound:
+                events = []
```
Comment (purplesmoke05) [inline batch/indexer_Position_Bond.py:194]: Fix handling exception as narrow as possible.

File: batch/indexer_Position_Bond.py [hunk 5]
```diff
@@ -192,7 +209,7 @@ def __sync_transfer(self, db_session: Session, block_from: int, block_to: int):
                                 exchange_commitment=_exchange_commitment
                             )
             except Exception as e:
-                LOG.exception(e)
+                raise e
```
Comment (purplesmoke05) [inline batch/indexer_Position_Bond.py:212]: If we get exceptions excluding ABIEventNotFound, processor will raise it.

File: batch/indexer_Position_Bond.py [hunk 6]
```diff
@@ -709,7 +774,18 @@ def __sink_on_position(db_session: Session,
...
+    while initial_synced_completed:
+        try:
+            processor.initial_sync()
+            LOG.debug("Initial sync is processed successfully")
+            initial_synced_completed = True
+        except Exception:
+            LOG.exception("Initial sync failed")
+
+        time.sleep(10)
+
```
[the remark this answers is not in the document]
Comment (purplesmoke05) [reply batch/indexer_Position_Bond.py:788]: I’m going to apply this change only on v22.6.

File: tests/batch/indexer_Position_Coupon_test.py [hunk 7]
```diff
@@ -437,3 +476,137 @@ def test_error_1(self, processor, shared_contract, session):
...
+            processor.sync_new_logs()
+
+        # Assertion
+        _position_list = session.query(IDXPosition).order_by(IDXPosition.created).all()
+        assert len(_position_list) == 0
+        # Latest_block is NOT incremented in "sync_new_logs" process.
+        assert processor.latest_block == block_number_bf
+
+    # <Error_3>: ServiceUnavailable occurs and is handled in mainloop.
+    def test_error_3(self, main_func, shared_contract, session, caplog):
```
Comment (purplesmoke05) [inline tests/batch/indexer_Position_Coupon_test.py:597]: When some exceptions occur in “sync_new_logs”(test_error_3), batch will output logs like below.

```
[2022-05-17 11:28:57 +0900] [INDEXER-POSITION-BOND] [90226] [INFO] Service started successfully
[2022-05-17 11:28:58 +0900] [INDEXER-POSITION-BOND] [90226] [ERROR] An exception occurred during event synchronization
Traceback (most recent call last):
  ~StackTrace
app.errors.ServiceUnavailable
[2022-05-17 11:28:58 +0900] [INDEXER-POSITION-BOND] [90226] [WARNING] An external service was unavailable
```

**Review by YoshihitoAso — APPROVED** [review]

**Pull request merged**